File: jenfengchang/463.cpp
class Solution {
public:
    int islandPerimeter(vector<vector<int>>& grid) {
        if(grid.size()==0 || grid[0].size()==0) {
            return 0;
        }
        int res = 0;
        const int n = grid.size();
        const int m = grid[0].size();
        for(int i = 0; i<n; i++) {
            for(int j = 0; j<m; j++) {
                if(grid[i][j]==1) {
                    int tmp = 0;
                    if(i>0 && grid[i-1][j]) {
                        tmp++;
                    }
                    if(j>0 && grid[i][j-1]) {
                        tmp++;
                    }
                    if((i+1)!=n && grid[i+1][j]) {
                        tmp++;
                    }
                    if((j+1)!=m && grid[i][j+1]) {
                        tmp++;
                    }
                    res += (4-tmp);
                }
            }
        }
        return res;
    }
};

File: zemiao/463.py
class Solution(object):
    def islandPerimeter(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        if not grid:
            return 0
        m, n, ans = len(grid), len(grid[0]), 0
        seen = [[0]*n for _ in range(m)]
                
        for a in range(m):
            for b in range(n):
                if seen[a][b] or grid[a][b] == 0:
                    continue
                seen[a][b] = 1

                ans += 4
                if a-1 >= 0:
                    ans -= 1 if grid[a-1][b]==1 else 0
                if a+1 < m:
                    ans -= 1 if grid[a+1][b]==1 else 0
                if b-1 >= 0:
                    ans -= 1 if grid[a][b-1]==1 else 0
                if b+1 < n:
                    ans -= 1 if grid[a][b+1]==1 else 0
        
        return ans
            